File: api/src/main/java/com/tersesystems/echopraxia/support/Utilities.java
package com.tersesystems.echopraxia.support;

import com.tersesystems.echopraxia.Field;
import java.lang.reflect.InvocationTargetException;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;
import org.jetbrains.annotations.NotNull;

/** Utilities for classes implementing a logger. */
public final class Utilities {

  @NotNull
  public static <T> T getNewInstance(@NotNull Class<T> newBuilderClass) {
    try {
      return newBuilderClass.getDeclaredConstructor().newInstance();
    } catch (NoSuchMethodException
        | SecurityException
        | InstantiationException
        | IllegalAccessException
        | InvocationTargetException e) {
      throw new IllegalStateException(e);
    }
  }

  @NotNull
  public static Function<Supplier<Map<String, String>>, Supplier<List<Field>>>
      getThreadContextFunction(@NotNull Function<Map<String, String>, List<Field>> f) {
    return mapSupplier -> () -> f.apply(mapSupplier.get());
  }

  public static <T> Supplier<T> memoize(Supplier<T> supplier) {
    return new MemoizingSupplier<>(supplier);
  }

  @NotNull
  static class MemoizingSupplier<T> implements Supplier<T> {
    final Supplier<T> delegate;
    transient volatile boolean initialized;
    transient T value;

    MemoizingSupplier(Supplier<T> delegate) {
      this.delegate = delegate;
    }

    @Override
    public T get() {
      if (!initialized) {
        synchronized (this) {
          if (!initialized) {
            T t = delegate.get();
            value = t;
            initialized = true;
            return t;
          }
        }
      }
      return value;
    }
  }
}

File: scala/api/src/main/scala/com/tersesystems/echopraxia/sapi/support/Utilities.scala
package com.tersesystems.echopraxia.sapi.support

import com.tersesystems.echopraxia.Field
import com.tersesystems.echopraxia.sapi.FieldBuilder

import java.util
import java.util.function.{Function, Supplier}
import scala.jdk.CollectionConverters._
import scala.util.control.NonFatal

object Utilities {

  def getNewInstance[T <: Field.Builder](newBuilderClass: Class[T]): T = {
    try {
      newBuilderClass.getDeclaredConstructor().newInstance()
    } catch {
      case NonFatal(e) =>
        throw new IllegalStateException(e)
    }
  }

  val getThreadContextFunction
      : Function[Supplier[util.Map[String, String]], Supplier[util.List[Field]]] =
    new Function[Supplier[util.Map[String, String]], Supplier[util.List[Field]]] {
      override def apply(
          mapSupplier: Supplier[util.Map[String, String]]
      ): Supplier[util.List[Field]] = {
        new Supplier[util.List[Field]]() {
          def buildFields(contextMap: util.Map[String, String]): util.List[Field] = {
            val list = new util.ArrayList[Field]();
            for (entry <- contextMap.entrySet().iterator.asScala) {
              list.add(FieldBuilder.keyValue(entry.getKey -> entry.getValue))
            }
            list
          }
          override def get(): util.List[Field] = buildFields(mapSupplier.get)
        }
      }
    }
}
